### ISSUE TITLE
Typing issue due to restrictive `__getattr__` of `nn.Module`

### ISSUE THREAD

**Issue by otaj** [pytorch/pytorch#81462]
### 🐛 Describe the bug

Hi,

let me preface this by saying I'm not 100% this is PyTorch bug and it could very well be issue of type checkers.

What happens: When you're using a subclass `nn.Module`, which defines extra methods, according to type checkers (tested mypy and Pylance), the extra method itself gets a type `Union[Tensor, Module]` from `__getattr__` method of `nn.Module`. The reason why this happens is that `__getattr__` of `nn.Module` is defined earlier than any of the extra methods of the subclass due to MRO resolution, therefore it kicks in sooner.

This issue manifested in PyTorch Lightning when we were trying to improve typing situation there (https://github.com/Lightning-AI/lightning/issues/13445). We stumbled upon this when the method `register_comm_hook` of `torch.nn.parallel.DistributedDataParallel` (which is a subclass of `nn.Module`) is not recognized as `Callable`, but as `Union[Tensor, Module]`

This leads to typing error: `"Tensor" not callable` when trying to invoke such method.
Please note, that the method can actually be invoked, but type checkers don't think so.

### Simple example

```python
from torch.nn.parallel import DistributedDataParallel
from typing_extensions import reveal_type

def register_hook(module: DistributedDataParallel, state: object, hook: Callable):
    reveal_type(module.register_comm_hook) # This outputs Union[Tensor, Module]
    module.register_comm_hook(state, hook) # This throws a typing error: "Tensor" not callable
```

### Versions

```
PyTorch version: 1.11.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 12.1.0
Clang version: Could not collect
CMake version: version 3.23.2
Libc version: glibc-2.35

Python version: 3.9.13 (main, Jun 17 2022, 10:41:03)  [GCC 12.1.0] (64-bit runtime)
Python platform: Linux-5.18.10-zen1-1-zen-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.64
GPU models and configuration: GPU 0: NVIDIA T1200 Laptop GPU
Nvidia driver version: 515.57
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.4.1
/usr/lib/libcudnn_adv_infer.so.8.4.1
/usr/lib/libcudnn_adv_train.so.8.4.1
/usr/lib/libcudnn_cnn_infer.so.8.4.1
/usr/lib/libcudnn_cnn_train.so.8.4.1
/usr/lib/libcudnn_ops_infer.so.8.4.1
/usr/lib/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.961
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.3
[pip3] torch==1.11.0
[pip3] torchmetrics==0.7.2
[pip3] torchtext==0.12.0
[pip3] torchvision==0.12.0
[conda] Could not collect
```

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ezyang @malfet @rgommers @xuzhao9 @gramster

**Comment by ezyang** [comment]
If someone can suggest a diff to fix this, I will happily review it. I don't know how to solve this problem, mypy just does not seem to have enough oomph for this case.

**Comment by otaj** [comment]
Hi @ezyang, I gotta say that I don't have an idea of what would be the appropriate fix for this.

However, I gotta say, it might not be a limitation of mypy, but python type system, since Pylance (or more specifically Pyright) was complaining as well with the same resolution of types.

**Comment by albanD** [comment]
That does look very surprising as `register_comm_hook` is a plain method on `DistributedDataParallel` and so shouldn't lead to any issue.
That being said, I also agree that our typing for `Module.__getattr__`, while right is overly restrictive (as you force all subclasses to only ever return Tensor/Module from this method).
I think a PR changing the return type of `__getattr__` to `Any` would be a good fix, WDYT @jbschlosser ?

Note on 3.11 this seems to be fixed as
```python
>>> reveal_type(DistributedDataParallel.register_comm_hook)
Runtime type is 'function'
<function DistributedDataParallel.register_comm_hook at 0x7fd532394050>
```

**Comment by otaj** [comment]
To be honest, my identification of the cause of the problem as `__getattr__` was a best guess estimate, but now that I am thinking about it more, I think it is Python not resolving types properly for multiple parent classes. The fact, that it resolves properly on 3.11 only supports that guess

**Comment by albanD** [comment]
In any case, users should be able to subclass a `Module` and have their custom `__getattr__` return anything they want. We don't actually rely on the fact that `__getattr__` only returns Tensors and Modules (AFAIK).

**Comment by jbschlosser** [comment]
> In any case, users should be able to subclass a `Module` and have their custom `__getattr__` return anything they want. We don't actually rely on the fact that `__getattr__` only returns Tensors and Modules (AFAIK).

Yep, I agree with this - let's try to update `__getattr__` return type to `Any`. Typing for `__setattr__` should probably be tweaked as well for consistency.

**Comment by jaketae** [comment]
I'll take this if anyone hasn't yet!

**Comment by albanD** [comment]
No-one that I know of! Feel free to take it!

**Comment by Xiao-Chenguang** [comment]
I have encountered the same problem when try out the official PyTorch tutorial for [transformer](https://pytorch.org/tutorials/beginner/transformer_tutorial.html). Specifically, the registered buffer **pe** is recognised as type of **Union[Tensor, Module]**, thus trigger pylance error of **"__getitem__" method not defined on type "Module"**.

### Simple example
```python
import math
import torch
from torch import nn, Tensor
    

class PositionalEncoding(nn.Module):

    def __init__(self, d_model: int, dropout: float = 0.1, max_len: int = 5000):
        super().__init__()
        self.dropout = nn.Dropout(p=dropout)

        position = torch.arange(max_len).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2) * (-math.log(10000.0) / d_model))
        pe = torch.zeros(max_len, 1, d_model)
        pe[:, 0, 0::2] = torch.sin(position * div_term)
        pe[:, 0, 1::2] = torch.cos(position * div_term)
        self.register_buffer('pe', pe)

    def forward(self, x: Tensor) -> Tensor:
        """
        Arguments:
            x: Tensor, shape ``[seq_len, batch_size, embedding_dim]``
        """
        x = x + self.pe[:x.size(0)]
        return self.dropout(x)
```

The error occurs in line `x = x + self.pe[:x.size(0)]` of the forward method.

**Comment by albanD** [comment]
This is now fixed in main by https://github.com/pytorch/pytorch/pull/104321. You can use the nightly to get this fix (or wait for the 2.1 release)
